mlir/lib/Conversion/PDLToPDLInterp/PDLToPDLInterp.cpp

```diff
@@ -364,6 +368,20 @@ Value PatternLowering::getValueAt(Block *&currentBlock, Position *pos) {
           loc, rawTypeAttr.cast<ArrayAttr>());
     break;
   }
+  case Predicates::ConstraintResultPos: {
+    // At this point in time the corresponding pdl.ApplyNativeConstraint op has
+    // been deleted and the new pdl_interp.ApplyConstraint has not been created
+    // yet. To enable use of results created by these operations we build a
+    // placeholder value that will be replaced when the actual
+    // pdl_interp.ApplyConstraint operation is created.
```

Comment:
I don't know enough about `PDLInterp`, could you give me an example when this is relevant?

Reply:
From my description on top:
> This is required in all scenarios where a native constraint returns a result that is used on the rhs
because said result has to be an input the the pdl_interp rewriter function that is created for the pattern.
However, the call to this function is always generated before the pdl_interp.apply_constraint operation is created.

An example for the output of this pass is the following:
PDL input:
```MLIR
module {
  pdl.pattern @simple : benefit(1) {
    %0 = operands
    %1 = operation "test.op"(%0 : !pdl.range<value>) 
    %attr = apply_native_constraint "NativeConstraint"(%1 : !pdl.operation) : !pdl.attribute 
    rewrite %1 {
      %3 = operation "test.success"  {"someAttr" = %attr} {"numRegions" = 1}
      replace %1 with %3
    }
  }
}
```
The PDL_interp output is the following:
```MLIR
module {
  pdl_interp.func @matcher(%arg0: !pdl.operation) {
    pdl_interp.check_operation_name of %arg0 is "test.op" -> ^bb2, ^bb1
  ^bb1:  // 4 preds: ^bb0, ^bb2, ^bb3, ^bb4
    pdl_interp.finalize
  ^bb2:  // pred: ^bb0
    pdl_interp.check_result_count of %arg0 is 0 -> ^bb3, ^bb1
  ^bb3:  // pred: ^bb2
    %0 = pdl_interp.apply_constraint "NativeConstraint"(%arg0 : !pdl.operation) : !pdl.attribute -> ^bb4, ^bb1
  ^bb4:  // pred: ^bb3
    pdl_interp.record_match @rewriters::@pdl_generated_rewriter(%0, %arg0 : !pdl.attribute, !pdl.operation) : benefit(1), generatedOps(["test.success"]), loc([%arg0]), root("test.op") -> ^bb1
  }
  module @rewriters {
    pdl_interp.func @pdl_generated_rewriter(%arg0: !pdl.attribute, %arg1: !pdl.operation) {
      %0 = pdl_interp.get_results of %arg1 : !pdl.range<value>
      %1 = pdl_interp.get_value_type of %0 : !pdl.range<type>
      %2 = pdl_interp.create_operation "test.success" {"someAttr" = %arg0}  -> (%1 : !pdl.range<type>) {numRegions = 1 : i64}
      pdl_interp.erase %arg1
      pdl_interp.finalize
    }
  }
}
```

The `pdl_interp.record_match` operation and the block it is in is not the last thing this pass generates for a matcher. This is done so the operation that is the last check before a successful match can have this block (^bb4) as successor (e.g. `%0 = pdl_interp.apply_constraint` here). But the `pdl_interp.record_match` needs the result of `%0 = pdl_interp.apply_constraint` as operand.
So there is a cyclic dependency here, which we solve using the placeholder.

In short: 
- `%0 = pdl_interp.apply_constraint` needs the success branch as successor.
- `pdl_interp.record_match` needs the result `%0` as operand